### PR TITLE
BUG: Correctly use `axis` in `scipy.stats.tmean`

### LINKED ISSUE #12106: `axis` option for `stats.tmean` do not appear to be working as expected
My issue is about (see title)

#### Reproducing code example:

```
import scipy as sp
import numpy as np

r = np.array([[1, 1, 1], [2, 2, 2], [3, 3, 3]])

print(sp.stats.tmean(r, axis=1))
print(sp.stats.tstd(r, axis=1))
```
Output:
```
2.0
[0. 0. 0.]
```

Expectation:
```
[1.0, 2.0, 3.0]
[0., 0., 0.]
```

#### Scipy/Numpy/Python version information:
```
1.4.1 1.18.1 sys.version_info(major=3, minor=7, micro=6, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by siddhantwahal**
Closes #12106

Fixes `scipy.stats.tmean` so that it correctly uses the `axis` keyword.

(_Please check that this addresses gh-9770, too_ - @mdhaber)

### CODE REVIEW

File: scipy/stats/stats.py [hunk 1]
```diff
@@ -558,11 +558,12 @@ def tmean(a, limits=None, inclusive=(True, True), axis=None):
 
     """
     a = asarray(a)
+    a = a.astype(float)
```
Comment (peterbell10) [inline scipy/stats/stats.py:561]: You can potentially avoid a copy here by passing `copy=false` to `astype`, or even better:
```suggestion
    a = asarray(a, dtype=np.float64)
```

File: scipy/stats/stats.py [hunk 2]
```diff
@@ -558,11 +558,12 @@ def tmean(a, limits=None, inclusive=(True, True), axis=None):
...
     if limits is None:
-        return np.mean(a, None)
-
-    am = _mask_to_limits(a.ravel(), limits, inclusive)
-    return am.mean(axis=axis)
+        return np.mean(a, axis)
+    am = _mask_to_limits(a, limits, inclusive)
+    amnan = am.filled(fill_value=np.nan)
+    return np.nanmean(amnan, axis=axis)
```
Comment (peterbell10) [inline scipy/stats/stats.py:566]: Are you sure the masked array mean is ignoring the `axis` argument. If so, that's a NumPy bug. Note that the example from the issue doesn't include limits, so never hits this code.

Comment (siddhantwahal) [reply scipy/stats/stats.py:566]: Yes, although the original issue doesn't mention this, it leads to an `AxisError` on `master`:

```
>>> r = np.array([[1, 1, 1], [2, 2, 2], [3, 3, 3]])
>>> stats.tmean(r, limits=(1, 2), axis=1)

----> 1 stats.tmean(r, limits=(1,2), axis=1)
~/scipy/scipy/stats/stats.py in tmean(a, limits, inclusive, axis)                                                                                                                   563
    564     am = _mask_to_limits(a.ravel(), limits, inclusive)                                                                                                                  --> 565     return am.mean(axis=axis)
    566                                                                                                                                                                             567
<snip snip>
AxisError: axis 1 is out of bounds for array of dimension 1
```

The fix I introduced produces the correct answer: `array([ 1.,  2., nan])`. 

I'm pretty sure the culprit is [here](https://github.com/scipy/scipy/blob/8040f6706acf43f4cbdc4705fe507cd459a60fdc/scipy/stats/stats.py#L564), where the input array is raveled. The `axis` keyword becomes meaningless then.

Comment (peterbell10) [reply scipy/stats/stats.py:565]: Yes, `ravel` would get rid of the axes. But you should be able to use `.mean(axis)` on the results without `ravel`.

Comment (siddhantwahal) [reply scipy/stats/stats.py:565]: Yes, that will work, although it'll then return a `MaskedArray`. I did it this way for consistency with `scipy.stats.tvar`.

Comment (siddhantwahal) [reply scipy/stats/stats.py:565]: There's already machinery to do this in `scipy.stats.mstats`. For example, `scipy.stats.mstats_basics.tmean` does the exact same thing:

```
>>> mstats.tmean(r, limits=(1,2), axis=1)
masked_array(data=[1.0, 2.0, --],
             mask=[False, False,  True],
       fill_value=1e+20)
```

I'm not sure, should we have two functions that do the exact same thing?

Comment (peterbell10) [reply scipy/stats/stats.py:565]: Ah, I see. This is slightly weird in that it changes the treatment of input nans, but I suppose being consistent with `tvar` is still an improvement.

The other option would be to access the `.data` attribute on the masked array when an axis is passed in.

Comment (siddhantwahal) [reply scipy/stats/stats.py:565]: Hmm, good point. How about something like this?
```suggestion
    am = _mask_to_limits(a, limits, inclusive)
    return am.mean(axis=axis).filled(fill_value=np.nan)
```

This propagates input nans to the output -- nans are only added and never disappear.

Comment (peterbell10) [reply scipy/stats/stats.py:565]: That seems reasonable, although I don't think it works for `axis=None` where `mean` returns a scalar. You would need something like
```python
return am.mean() if axis is None else am.mean(axis).filled(fill_value=np.nan)
```

Comment (peterbell10) [reply scipy/stats/stats.py:565]: Hrm, actually even that's not quite right because you can have `axis=0` for a 1d array and the result is still a scalar.

Comment (peterbell10) [reply scipy/stats/stats.py:565]: I think this version is safe:
```python
mean = np.ma.filled(am.mean(axis=axis), fill_value=np.nan)
return mean if mean.ndim > 0 else mean.item()
```

Comment (siddhantwahal) [reply scipy/stats/stats.py]: Thanks for the suggestions @peterbell10, I've incorporated then in the PR.

File: scipy/stats/tests/test_stats.py [hunk 3]
```diff
@@ -62,6 +62,44 @@ def test_tmean(self):
...
+        y_true = [14, 11.5, 9, 10, 11, 12, 13]
+        assert_array_almost_equal(y, y_true, decimal=8)
+
+        y = stats.tmean(x_2d, limits=(2, 21), inclusive=(True, False), axis=0)
+        y_true = [10.5, 11.5, 9, 10, 11, 12, 13]
+        assert_array_almost_equal(y, y_true, decimal=8)
+
+        x_2d_with_nan = np.array(x_2d)
+        x_2d_with_nan[-1, -3:] = np.nan
+        y = stats.tmean(x_2d_with_nan, limits=(1,13), axis=0)
```
Comment (peterbell10) [inline scipy/stats/tests/test_stats.py:88]: Lint is failing:
```suggestion
        y = stats.tmean(x_2d_with_nan, limits=(1, 13), axis=0)
```

**Review by peterbell10 — APPROVED** [review]

**Comment by siddhantwahal** [comment]
Thanks a lot for reviewing @peterbell10 :smile:

File: scipy/stats/stats.py [hunk 4]
```diff
@@ -557,12 +557,12 @@ def tmean(a, limits=None, inclusive=(True, True), axis=None):
     10.0
 
     """
-    a = asarray(a)
+    a = asarray(a, dtype=np.float64)
```
Comment (larsoner) [inline scipy/stats/stats.py:560]: Is this necessary? Shouldn't it work on at least `float32` and maybe also `float16` and not need to convert?

**Review by larsoner — COMMENTED** [review]
Other than a dtype conversion question, LGTM

**Comment by larsoner** [comment]
Will need a rebase to fix a merge conflict

File: scipy/stats/stats.py (hunk 4 above)
Comment (siddhantwahal) [reply scipy/stats/stats.py]: You're right. `np.mean` and `am.mean` cast where necessary.

**Comment by siddhantwahal** [comment]
I just realized that the NaN policy of this `tmean` implementation is inconsistent with other functions in the module. Some, like `tmin, tmax`, explicitly ask the user for a `nan_policy`, `tvar` doesn't, while this `tmean` forcibly propagates input NaNs. 

Is it worthwhile to add a `nan_policy` argument to `tmean` in this PR?

File: THANKS.txt [hunk 5]
```diff
@@ -236,7 +236,9 @@ Sam Wallan for scipy.linalg.lapack enhancements
 Richard Weiss for a bug fix in scipy.optimize._differentialevolution.py.
 Luigi F. Cruz for adding time/frequency domain option to scipy.signal.resample.
 Wesley Alves for improvements to scipy.stats.jarque_bera and scipy.stats.shapiro
+<<<<<<< HEAD
```
Comment (larsoner) [inline THANKS.txt:239]: rebase error

**Review by larsoner — COMMENTED** [review]
> Is it worthwhile to add a nan_policy argument to tmean in this PR?

My vote would be to do this in a separate PR

Other than a rebase problem, this looks reasonable to me

File: THANKS.txt (hunk 5 above)
Comment (siddhantwahal) [reply THANKS.txt:239]: Aargh, don't know how I missed that, sorry.

**Comment by WarrenWeckesser** [comment]
An invariant of `tmean` is that `tmean(a, limits=(None, None)) == tmean(a)`.  But there is a bug in the handling of `inf`, in both this PR and in master.  In the master branch, we have, for example,

```
In [7]: a = np.array([1, 2, 3, np.inf])

In [8]: tmean(a)
Out[8]: inf

In [9]: tmean(a, limits=(None, None))
Out[9]: masked
```
With this pull request, we get
```
In [51]: a = np.array([1, 2, 3, np.inf])

In [52]: tmean(a)
Out[52]: inf

In [53]: tmean(a, limits=(None, None))
Out[53]: nan
```
The problem is the use of masked arrays for intermediate calculations. The masked arrays automatically mask `inf` values, which is not the desired behavior here.  Could you take a look at this issue, too, along with the `axis` problem?

**Comment by WarrenWeckesser** [comment]
The problem that I noted above is a slight change to a pre-existing bug, so it would be OK to put off fixing that to a separate follow-up pull request.

Here's a problem related to handling of the `axis` that should probably be addressed here.

Consider what happens with a 2-d array where some slices have values within the given limits and some don't:

```
In [21]: b
Out[21]: 
array([[1, 2, 3, 4],
       [5, 6, 7, 8]])

In [22]: tmean(b, limits=(0, 4), axis=1)
Out[22]: array([2.5, nan])
```
That's reasonable behavior: given an empty data set, return `nan` for the mean.  No warning is generated.

Given that, I expect `tmean(b[1], limits(0, 4))` to return `nan` with no warning.  But instead, it raises an exception:

```
In [23]: tmean(b[1], limits=(0, 4))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-23-c09c4c137566> in <module>
----> 1 tmean(b[1], limits=(0, 4))

~/mc37scipy/lib/python3.7/site-packages/scipy-1.5.0.dev0+fffefae-py3.7-macosx-10.9-x86_64.egg/scipy/stats/stats.py in tmean(a, limits, inclusive, axis)
    561     if limits is None:
    562         return np.mean(a, axis)
--> 563     am = _mask_to_limits(a, limits, inclusive)
    564     mean = np.ma.filled(am.mean(axis=axis), fill_value=np.nan)
    565     return mean if mean.ndim > 0 else mean.item()

~/mc37scipy/lib/python3.7/site-packages/scipy-1.5.0.dev0+fffefae-py3.7-macosx-10.9-x86_64.egg/scipy/stats/stats.py in _mask_to_limits(a, limits, inclusive)
    511 
    512     if am.count() == 0:
--> 513         raise ValueError("No array values within given limits")
    514 
    515     return am

ValueError: No array values within given limits
```

The same exception is raised with a call such as `tmean(b, limits=(0, 0.1), axis=1)`.

What does `numpy.mean` (currently) do with an empty input?  It generates a warning and returns `nan`:

```
In [24]: np.mean([])
/Users/warren/mc37scipy/lib/python3.7/site-packages/numpy/core/fromnumeric.py:3335: RuntimeWarning: Mean of empty slice.
  out=out, **kwargs)
/Users/warren/mc37scipy/lib/python3.7/site-packages/numpy/core/_methods.py:161: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
Out[24]: nan
```
So I think `tmean` should not raise an exception with an empty input, and instead it should return `nan`.  As far as generating a warning, I could go either way.  If we clearly state in the docstring that empty input results in `nan`, I'm fine with not generating a warning.

**Comment by siddhantwahal** [comment]
To summarize additional issues that we've discovered in this module `scipy.stats`:

1. Inconsistent NaN policies across trimmed methods (`tmean, tvar, tstd, tmin, tmax` etc.)
2. Incorrect handling of `np.inf`
3. Incorrectly handling situations where the intermediate masked array is empty. This is a module level bug because nearly all methods in the module use `_mask_to_limits`. For instance, `stats.tvar(np.array([1, 2, 3, 4]), limits=(0, 0.1))` raises the same `ValueError`. 

Another module that implements very similar functionality is `scipy.stats.mstats`. The only difference is that it returns masked arrays instead of `ndarray`s. This module behaves correctly in some situations:

```python
>>> b = np.array([1, 2, 3, 4])
>>> mstats.tmean(b, limits=(0, 0.1))
masked
```

But, as expected, cannot handle `np.inf`:
```python
>>> mstats.tmean(np.array[1., 2., np.inf])
masked
```

Interestingly, 
```python
>>> mstats.tvar(np.array([1, 2, 3, 4]), limits=(0, 0.1))

~/scipy/scipy/stats/mstats_basic.py in tvar(a, limits, inclusive, axis, ddof)
   1875         n = (~a.mask).sum()  # todo: better way to do that?
   1876         return np.ma.var(a) * n/(n-1.)
-> 1877     am = _mask_to_limits(a, limits=limits, inclusive=inclusive)
   1878
   1879     return np.ma.var(am, axis=axis, ddof=ddof)

~/scipy/scipy/stats/mstats_basic.py in _mask_to_limits(a, limits, inclusive)
   1780
   1781     if am.count() == 0:
-> 1782         raise ValueError("No array values within given limits")
   1783
   1784     return am

ValueError: No array values within given limits
```
Our old friend `_mask_to_limits` also lives here. For some reason, `mstats.tmean` uses a different function, `trima`, to create the intermediate masked matrix, than `mstats.tvar`, which uses `_mask_to_limits`. 

Both modules do very similar things, and fail in very similar conditions. Should both be fixed? Or should we fix `mstats`, and reuse that logic in `stats`?

**Comment by mdhaber** [comment]
@WarrenWeckesser what do you think? Fix `mstats` and call it in `stats`?

**Comment by mdhaber** [comment]
Thanks @siddhantwahal. I think we'll fix this as part of gh-14651 using the decorator. That will also add `nan_policy` and support for `axis` tuples.

**Comment by mdhaber** [comment]
This PR makes the function better than it was.

> So I think `tmean` should not raise an exception with an empty input, and instead it should return nan

This is all over `scipy.stats`, and not just because of `_mask_to_limits`. It is not new because of this PR, so it does not need to be fixed here. In fact, I have not been fixing this sort of thing as part of the gh-14651 effort. The decorator could easily fix this behavior, but I actually took steps to prevent the decorator from fixing these because users could be looking for the `ValueError`s at this point. This sort of thing needs to be changed, but I think it needs some outside discussion about the general problem before we fix just this function here.

> The problem that I noted above is a slight change to a pre-existing bug, so it would be OK to put off fixing that to a separate follow-up pull request.

Agreed. `nan` is better than `masked`, in this case. 

> what do you think? Fix mstats and call it in stats?

That can be done separately. Part of the goal of gh-14651 is the unification of `stats` and `mstats`; it does not need to be done here.

I'll wait for CI to finish successfully and then merge.

**Comment by mdhaber** [comment]
@tylerjereddy if it's not too much trouble, can we backport this one? It's not essential, but this does fix a bug.

File: scipy/stats/_stats_py.py [hunk 6]
```diff
@@ -559,10 +559,10 @@ def tmean(a, limits=None, inclusive=(True, True), axis=None):
     """
     a = asarray(a)
     if limits is None:
-        return np.mean(a, None)
-
-    am = _mask_to_limits(a.ravel(), limits, inclusive)
-    return am.mean(axis=axis)
+        return np.mean(a, axis)
+    am = _mask_to_limits(a, limits, inclusive)
+    mean = am.mean(axis=axis)
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:564]:
```suggestion
    mean = np.ma.filled(am.mean(axis=axis), fill_value=np.nan)
```

**Comment by mdhaber** [comment]
Thanks @siddhantwahal @peterbell10!

**Pull request merged**